[Distributed] Augment accessor to support calling generic distributed thunks

xedin opened the pull request:
Previously distributed target accessors and `executeDistributedTarget` machinery
supported only distributed methods with concrete arguments and result. 

This PR makes it possible to call generic distributed methods through their thunks by
decoding generic arguments (provided by the caller) and their conformance requirements,
and providing accessor with all the information necessary to form generic invocations. 
Accessor itself has been augmented to deal with generic parameters, substitutions 
and witness tables.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[IRGen] Record generic environment associated with accessible function

Without this piece of information it wouldn't be possible to form
a call to the recorded function at runtime if it's generic.

xedin committed:
[Distributed] Runtime: Implement a way to access generic env of distributed target

xedin committed:
[Distributed] Stage use of generic environment to decode target parameter/result types

xedin committed:
[Distributed] Augment parameter/result type decoded to support generic substitutions

xedin committed:
[Distributed] IRGen: Provide argument/result types to the accessor

xedin committed:
[Distributed] IRGen: Support loading both concrete and generic argument values

To load generic argument value, we need to retrieve argument type
from the argument types buffer, based on that load value size and
alignment, and only then fetch actual value from the argument buffer
as `swift.opaque *`.

xedin committed:
[Distributed] IRGen: Drop result type for accessor parameters

xedin committed:
[Distributed] Runtime/IRGen: Allow passing substitutions to distributed accessor

Accept an optional list of substitutions which are required to form
a call to a generic distributed thunk.

xedin committed:
[Distributed] IRGen: Add generic argument substitutions to argument explosion of thunk call

xedin committed:
[Distributed] Runtime: Add a way to fetch witness tables for given generic env/arguments

xedin committed:
[Distributed] ExecuteTarget: Add missing deallocation of substitutions buffer

xedin committed:
[Distributed] IRGen: Support argument with generic parameters passed directly

This is the case of e.g. `Array<T>` where argument is `Array<Int>`
we need to compute offset/size based on argument metadata but load
the argument using parameter type information.

xedin committed:
[Distributed] IRGen: Teach distributed thunk accessor to load witness tables (when needed)

ktoso commented on test/Distributed/Runtime/distributed_actor_remoteCall.swift:
```diff
@@ -255,6 +264,25 @@ func test() async throws {
+
+  // CHECK: ---> T = 42, type(of:) = Int
+  // CHECK-NEXT: ---> U = Ultimate Question!, type(of:) = String
+  // CHECK-NEXT: RETURN: ()
```
🥳

ktoso commented on stdlib/public/runtime/MetadataLookup.cpp:
```diff
@@ -2039,40 +2057,75 @@ swift_func_getParameterTypeInfo(
+  for (unsigned i = 0, n = witnessTables.size(); i != n; ++i)
+    tables[i] = const_cast<void *>(witnessTables[i]);
+
+  return {tables, static_cast<intptr_t>(witnessTables.size())};
```
cool, so no back and forth for the alloc and we just return :)

xedin replied:
Yeah, I wasn't sure whether that was correct but apparently it is.

ktoso commented on stdlib/public/Distributed/DistributedActorSystem.swift:
```diff
@@ -207,11 +248,11 @@ extension DistributedActorSystem {
+      argumentTypes.reserveCapacity(Int(decodedNum))
+      for argumentType in argumentTypesBuffer {
+        argumentTypes.append(argumentType)
       }
```
thx for the rename

ktoso commented on stdlib/public/Distributed/DistributedActorSystem.swift:
```diff
@@ -168,6 +168,47 @@ extension DistributedActorSystem {
+          message: "Generic substitutions \(subs) do not satisfy generic requirements of \(mangledTargetName)")
+      }
+    }
+
```
Sweet 🎉

ktoso reviewed (approved):
Looking good :) Not really in depth in the low level IR pieces, but all looks good to me... Thanks!

xedin committed:
[Distributed] Runtime: Move substitutions/witness tables deallocation to the outer context

Fixes a bug where substitutions and witness tables buffers are going
to be deallocated at the end of the `if` block before calling
`_executeDistributedTarget`.

xedin committed:
[Distributed] NFC: Add more test-cases for generic distributed methods

xedin committed:
[Distributed] NFC: Adjust test-cases to reflect updates in accessor interface

xedin committed:
[Distributed] NFC: Add generic argument tests for 64 and 32 bit accessors

aschwaighofer commented on lib/IRGen/GenDistributed.cpp:
```diff
@@ -113,6 +150,11 @@ static CanSILFunctionType getAccessorType(IRGenModule &IGM,
 
+  auto getUIntParameter = [&]() {
+    return SILParameterInfo(Context.getUIntType()->getCanonicalType(),
+                            ParameterConvention::Direct_Guaranteed);
```
Technically, `Direct_Unowned`?

xedin replied:
I picked guaranteed because the comment said that `caller guarantees its validity for the entirety of the call.` but I can change it to `Direct_Unowned` if that's better since I'm not privy to all of the nuances that differentiate the two.

aschwaighofer replied:
We use `Unowned` for trivial types.

```
  /// This argument is passed directly.  Its type may be trivial, or it may     
  /// simply be that the callee is not responsible for destroying it.  Its      
  /// validity is guaranteed only at the instant the call begins.               
  Direct_Unowned,                                                               
                                                                                
  /// This argument is passed directly.  Its type is non-trivial, and the caller
  /// guarantees its validity for the entirety of the call.                     
  Direct_Guaranteed,   
  ```

xedin replied:
So since all these are either pointers or i{{64|32}} I should have used `Unowned` but if I want to e.g. pass a reference to class or struct it should be `Guaranteed`?

xedin replied:
Updated!

xedin committed:
[Distributed] Switch accessor parameter convention to `Direct_Unowned` since all of them are trivial

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@compnerd Could you please help me understand the issue on Windows? The compiler says this:

```
C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\stdlib\public\Concurrency\Actor.cpp(1714,7): error: dllimport cannot be applied to non-inline function definition
void *swift_distributed_get_generic_environment(const char *targetNameStart,
      ^
```

But I'm not sure what is the difference between this function and all the rest that are declared in DistributedMetadata.swift...

xedin commented:
For some reason 32-bit WatchOS still fails in CI but it actually runs just fine on my machine, might be something to do with OS/similar version in CI environment...

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please test Linux platform

compnerd commented:
@xedin - I think that is the problem in its own right - "this matches the other declarations" ;-)

If you look at the error, it tells you what is going wrong:
```
C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\stdlib\public\Concurrency\Actor.cpp(1714,7): error: dllimport cannot be applied to non-inline function definition
void *swift_distributed_get_generic_environment(const char *targetNameStart,
      ^
```

The function is being marked as `dllimport`, meaning that it believes that it is coming from a different module than the one being built.  From the command line, we can ascertain that the module being built is `swift_Concurrency`.

Looking at the diff we see:

```
SWIFT_CC(swift)
 SWIFT_RUNTIME_STDLIB_SPI
 void *swift_distributed_get_generic_environment(const char *targetNameStart,
                                                 size_t targetNameLength) {
```

The problem is that definition.  It is *NOT* a runtime standard library SPI - it doesn't go into the standard library, it is going into a different module (Concurrency).  That should be marked as `SWIFT_EXPORT_FROM(swift_Concurrency)` which should resolve the issue.

xedin commented:
For posterity - we have discussed the issue and the best course of action to is move code out of Actor.cpp which I'm going to do now.

xedin committed:
[Distributed] Move some of the swift_distributed_* accessors to a dedicated file

Add a new `DistributedActor.cpp` and move some of the logic related to
swift_distributed_execute_target there to avoid exporting it from Concurrency.

xedin commented:
@swift-ci please test Windows platform

xedin commented:
@swift-ci please test Windows platform

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please clean test Linux platform

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please test Linux platform

xedin committed:
[Distributed] Add stdlib include to give access to llvm/ADT

xedin committed:
[Distributed] NFC: Adjust generic argument test-case to match `Array` representation on Linux correctly

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please test Windows platform
